### PR TITLE
Fix lexing hash directives and raise errors (2nd attempt) - several cases

### LINKED ISSUE #3487: Invalid directives are not a syntax error, unless in specific places
Using a made-up directive (or a typo) will not yield a syntax warning. I think it should. Though some typos are caught.

### Useful FS1169, but incorrect:

    #iff FALSE
          something illegal
    #endif

gives: 

> FS1169: #if directive should be immediately followed by an identifier

Though, from a syntax standpoint, the directive keyword here is `iff`, not `if`, so it should be recognized as an invalid directive, rather than a misuse of the `#if` .

### Error, but unexpected

    #foo 12

gives:

>  FS0010: Unexpected integer literal in implementation file. Expected incomplete structured construct at or before this point or other token.

Probably not the text you'd expect, and certainly not where you might expect it, as it is `#foo` that's illegal, `12` is irrelevant:

image

### No error at all, but totally illegal

    #foo "test"

### Warning, but unexpected

    module X =
        #foo "test"
        ()

gives this warning:

> FS0236: Directives inside modules are ignored

But it is illegal syntax and should give an error.

## Expected behavior

All scenarios above should give an error on the the start of the false, non-existent directive.

## Actual behavior

Errors, warnings, or even no errors where it is clearly illegal syntax

## Known workarounds

Use your brains, don't rely on the compiler to check for valid syntax ;).

## Related information 

Seen on current and previous versions of F#, so this is not a new bug.

In theory, failing capturing these syntax errors could be exploited by a preprocessor to do something useful with these directives. That means that fixing this is potentially backward-incompatible if users have indeed exploited this bug. Still, I would suggest to fix it, or turn it into a proper feature (for instance with a hook to do that preprocessing in a neat way).

Do note, however, that the F# 4.0 spec, section 12.4, shows in the syntax that there must be a space between the directive and the string or strings (and it contains a typo, suggesting there is a space between `#`  and the identifier). Also, the specification only lists what are valid directive, suggesting anything else should be invalid (i.e., `#iff`, `#foo` from the examples above).

### LINKED ISSUE #9787: FS0000 warning (not assigned) for #light or #indent being misplaced
The following warning has no error number number assigned (and text is wrong and not internationalized, I pick that up in #9779)

image

> test.fs(3,1): warning FS0000: #light should only occur as the first non-comment text in an F# source file

**Repro steps**

```
#light 
module Foo 
#indent "on"   // raises FS0000 warning
```

**Expected behavior**

Proper warning with a warning number, the token is `#indent`, not `#light`

**Actual behavior**

See screenshot: a warning without a warning number.

## Proposed fix

I've fixed the text already, but there's two more things that need to be done here:

1. It should be put inside `FSComp.SR` and translated
2. A number should be assigned. 

I can do (1), and I believe then the translations will be picked up, but I don't know how to do (2).

### PR DESCRIPTION
**Pull request by abelbraaksma**
Fixes #3487.
Fixes #9787.

_[continued from #9779, because somehow Github didn't understand the diff-view anymore, showing 100s unrelated non-existing changes. Close/recreate PR with exactly the same changes seemed to fix this]_

This resurrects the work done in https://github.com/dotnet/fsharp/pull/5498 by @misonijnik . In short, it's an attempt to make sure that:

```f#
#iff FALSE   // raises wrong error, this should fix that (should be invalid token error)
#foo "test" // doesn't raise error at all, but should (invalid token)
#foo 42   // raises wrong error, this should fix that (should be invalid token error)

module X =
    #foo 42   // shows warning, but should be error (invalid token)
```

Some tests were failing on the original PR, I merged it with latest master and marked this WIP, as I want to see what those errors were, though I don't expect to work long on it, the change is pretty benign ;).

## List of fixes (will be updated if necessary)

* [ ] Fix wrong hash-token parsing (partially done)
    * with the original change by @misonijnik, `#` is now only recognized and leading to tokens if it is valid. This should be updated to include invalid items and the proper error and range for the lang service.

* [x] Allow a type to start with the letters "if"

    **Previously**
    image

    **Implemented:**
    image

* [x] Fix wrong error for use of `#indent` (text mentions '#light', instead of '#indent') (#9787):
image

    **Implemented, whole token shown:**
    image

* [x] Fix wrong error for use of `#light` (#9787): 
image

    **Implemented:**
    image

* [ ] Fix wrong error when using wrong argument with `#light`:
image
* [ ] Improve `#light` when token is wrong:
image

    After @misonijnik's change (error and range still to improve):
    image

* [ ] Fix wrong error with wrong token with `#if`:
image

    After @misonijnik's change (error and range still to improve):
    image

 * [x] This fsharpqa test explicitly tests for leniency with hash-types starting with `#light`: `Conformance\TypesAndTypeConstraints\TypeParameterDefinitions (HashConstraint02.fs)`.
    ```
    // Regression test for FSHARP1.0:1419
    // Tokens beginning with # should not match greedily with directives
    // The only case where we are still a bit confused is #light: for this reason the code
    // below compiles just fine (it would not work if I replace #light with #r for example)
    #light
    
    type light_() = class
                    end
    
    let t = new light_()
    
    let t5 (x : #light_) = x       // this is now valid, and does not throw a warning
    
    let r1 = t5 1.0         // this is now INVALID! (as it should)
    let r2 = t5 t
    
    exit 0
    ```
    **Current situation**
    image

    **New situation**
    image

* [ ] Fix wrong parsing of `light` when used as hash-constraint:

    **Current situation**
    image

Generally: when `#` is followed by an unrecognized token, I would expect an error that says that the token is not recognized.

TODO:

* [x] Fix failing build
* [ ] Verify & improve errors / warnings
* [ ] add tests
* [ ] Review / improve code changes

### CODE REVIEW

**Comment by cartermp** [comment]
@abelbraaksma for the sake of incrementality, how would you feel if we just reviewed and aimed to merge what is here already?

**Comment by abelbraaksma** [comment]
@cartermp, that's probably a good idea. Let me merge and fix over the weekend so that this can be reviewed.

**Review by cartermp — APPROVED** [review]
Thanks! I think this looks good.

**Review by KevinRansom — APPROVED** [review]
Looks, good, but what do you want to do with the todo?

File: src/fsharp/lex.fsl [hunk 1]
```diff
@@ -895,8 +895,10 @@ rule token args skip = parse
  | "#light" anywhite* 
  | ("#indent" | "#light") anywhite+ "\"on\""
    { if args.lightStatus.ExplicitlySet && args.lightStatus.WarnOnMultipleTokens then 
-         warning(Error((0,"#light should only occur as the first non-comment text in an F# source file"), lexbuf.LexemeRange))
-     // TODO unreachable error above, I think? - brianmcn 
+         let s = lexeme lexbuf
+         warning(Error((0, sprintf "%s should only be set once in an F# source file." s), lexbuf.LexemeRange))
+         // TODO: where should this go? (abelb)
```
[the remark this answers is not in the document]
Comment (abelbraaksma) [reply src/fsharp/lex.fsl:900]: @KevinRansom I think I meant that all errors go in a resource file (iirc) and this line as it is is not localizable. When I wrote that I didn't know where to place the localizable string.

The other thing here is probably the error/warning number, which here is 0. I don't know what number to choose. It's been like that for a while, so I guess I could also fix it in a future pr. How are these numbers assigned? Just choose a free one?

[the remark this answers is not in the document]
Comment (abelbraaksma) [reply src/fsharp/lex.fsl:900]: Side effect of current code is, I guess, that you won't be able to disable this warning as it is written...

**Comment by cartermp** [comment]
I will merge this since the comment is benign.

> How are these numbers assigned? Just choose a free one?

Yep, you can create a new one in FSComp.txt and when you run `build` it will generate the necessary resource files.

**Comment by cartermp** [comment]
Thanks for the contribution! Feel free to follow up as you see fit.

**Comment by dsyme** [comment]
This one caused a build break on integration

```
/home/vsts/work/1/s/src/fsharp/lex.fsl(984,50): error FS0039: The type 'LexBuffer<Char>' does not define the field, constructor or member 'Lexeme'. Maybe you want one of the following:   LexemeView   LexemeChar   LexemeRange   LexemeString   LexemeContains [/home/vsts/work/1/s/src/fsharp/FSharp.Compiler.Service/FSharp.Compiler.Service.fsproj]
##[error]src/fsharp/lex.fsl(984,50): error FS0039: The type 'LexBuffer<Char>' does not define the field, constructor or member 'Lexeme'. Maybe you want one of the following:   LexemeView   LexemeChar   LexemeRange   LexemeString   LexemeContains
```

**Comment by dsyme** [comment]
Changing

   { let n = Array.IndexOf(lexbuf.Lexeme, '#')

to

   { let n = (lexeme lexbuf).IndexOf('#')

worked for me

I'll submit a PR

**Pull request merged**